### PR TITLE
Add attachments to `Hint`

### LINKED ISSUE #682: (PR created) Allow attachments to be bound to exception more easily
## Why I need this (very much!)

My code is like:

```dart
void main_function() {
  try {
    inner_function(); // indeed, functions calling functions calling functions... so the calling stack is deep
  } catch (e,s) {
    Sentry.captureException(e, s);
  }
}

void inner_function() {
  var my_data = ...;
  try {
    do_some_work_that_can_throw_exception(my_data);
  } catch (e,s) {
     Sentry.pleaseBindAttachmentToException(e, my_data); // <-- !!
     rethrow;
  }
}
```

Notice the `Sentry.pleaseBindAttachmentToException`, where I want to bind attachment to exception.

However, as we know, we cannot do it. I can create some code to bind `extra` to exception, because `extra` can be modified in EventProcessors (so I can manually create a static map holding some extras and bind to exception in my own EventProcessor). However, EventProcessors cannot modify or add attachments. So I have no way to do it :(

## Proposed solution

Say, add a field to `SentryOptions`, e.g. `void Function(Scope)? withScopeWhenCapture`. Then, in `Hub.captureException` and similar places, call that callback. This is very simple and I will implement it if you like.

related: https://github.com/getsentry/sentry-dart/issues/646

Hi @marandaneto could you please have a look (since this thread is strongly related to #646)? Thanks! And @ueman could you please also have a look? Thanks!

### PR DESCRIPTION
**Pull request by denrase**
## :scroll: Description

- Add attachments to `Hint`

## :bulb: Motivation and Context

Closes #682

## :green_heart: How did you test it?

- Added tests

## :pencil: Checklist

- [x] I reviewed submitted code
- [x] I added tests to verify changes
- [x] No new PII added or SDK only sends newly added PII if `sendDefaultPii` is enabled
- [ ] I updated the docs if needed
- [x] All tests passing
- [x] No breaking changes

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -13,6 +13,7 @@
 - Sync `connectionTimeout` and `readTimeout` to Android ([#1397](https://github.com/getsentry/sentry-dart/pull/1397))
 - Set User `name` and `geo` in native plugins ([#1393](https://github.com/getsentry/sentry-dart/pull/1393))
 - Add processor count to device info ([#1402](https://github.com/getsentry/sentry-dart/pull/1402))
+- Add attachments to `Hint` ([#1404](https://github.com/getsentry/sentry-dart/pull/1404))
```
Comment (marandaneto) [inline CHANGELOG.md:16]: Let's add a code snippet, similar to https://github.com/getsentry/sentry-dart/pull/1403
Let's make this a habit :)

File: dart/lib/src/sentry_client.dart [hunk 2]
```diff
@@ -323,6 +324,9 @@ class SentryClient {
     final attachments = scope?.attachments
         .where((element) => element.addToTransactions)
         .toList();
+
+    // TODO Do we need to add attahcments?
```
Comment (marandaneto) [inline dart/lib/src/sentry_client.dart]: Only if `addToTransactions`.

File: dart/lib/src/hint.dart [hunk 3]
```diff
@@ -20,15 +22,48 @@ import 'sentry_attachment/sentry_attachment.dart';
 ///   };
 /// }
 /// ```
+///
+/// The hint can also be used to add attachments to events.
```
Comment (marandaneto) [inline dart/lib/src/hint.dart]:
```suggestion
/// The[Hint] can also be used to add attachments to events.
```

File: dart/lib/src/sentry_client.dart (hunk 2 above)
Comment (denrase) [reply dart/lib/src/sentry_client.dart]: Was just a temp comment, as we don't pass `Hint` to `captureTransaction`.

File: CHANGELOG.md (hunk 1 above)
Comment (marandaneto) [inline CHANGELOG.md:16]:
```suggestion
- Add attachments to `Hint` ([#1404](https://github.com/getsentry/sentry-dart/pull/1404))

```dart
    final text = 'This event should not be sent happen in prod. Investigate.';
    final textAttachment = SentryAttachment.fromIntList(
      utf8.encode(text),
      'event_info.txt',
      contentType: 'text/plain',
    );
    await Sentry.captureException(
      exception,
      stackTrace: stacktrace,
      hint: Hint.withAttachment(textAttachment),
    );
```
```

Comment (marandaneto) [reply CHANGELOG.md:16]: Indentation is probably wrong and the backticks in suggestion does not work, but you got the idea.

Comment (denrase) [reply CHANGELOG.md:16]: pls check if this is ok, looks good in my local markdown editor

Comment (marandaneto) [reply CHANGELOG.md:16]: @denrase the snippet is fine, just remove the first 2 spaces of each line.

**Review by marandaneto — APPROVED** [review]

**Review by marandaneto — APPROVED** [review]

**Pull request merged**